air/src/proof/security.rs

```diff
@@ -0,0 +1,728 @@
+// Copyright (c) Facebook, Inc. and its affiliates.
+//
+// This source code is licensed under the MIT license found in the
+// LICENSE file in the root directory of this source tree.
+
+//! Contains helper structs and methods to estimate the security of STARK proofs.
+
+use core::cmp;
+
+use crate::ProofOptions;
+
+const GRINDING_CONTRIBUTION_FLOOR: u32 = 80;
+const MAX_PROXIMITY_PARAMETER: u64 = 1000;
+
+pub struct ConjecturedSecurity(u32);
+
+impl ConjecturedSecurity {
+    pub fn compute(
+        options: &ProofOptions,
+        base_field_bits: u32,
+        trace_domain_size: usize,
+        collision_resistance: u32,
+    ) -> Self {
+        // compute max security we can get for a given field size
+        let field_size = base_field_bits * options.field_extension().degree();
+        let field_security = field_size - (trace_domain_size * options.blowup_factor()).ilog2();
+
+        // compute security we get by executing multiple query rounds
+        let security_per_query = options.blowup_factor().ilog2();
+        let mut query_security = security_per_query * options.num_queries() as u32;
+
+        // include grinding factor contributions only for proofs adequate security
+        if query_security >= GRINDING_CONTRIBUTION_FLOOR {
+            query_security += options.grinding_factor();
+        }
+
+        Self(cmp::min(cmp::min(field_security, query_security) - 1, collision_resistance))
+    }
+
+    pub fn bits(&self) -> u32 {
+        self.0
+    }
+
+    pub fn is_at_least(&self, bits: u32) -> bool {
+        self.0 >= bits
+    }
+}
+
+pub struct ProvenSecurity {
+    unique_decoding: u32,
+    list_decoding: u32,
+}
+
+impl ProvenSecurity {
+    pub fn compute(
```

Comment:
Since this is a part of the public interface, let's add doc comments to all public methods of this struct.

Reply:
added